packages/jupyterlab-manager/src/semverregistry.ts

```diff
@@ -0,0 +1,151 @@
+      throw new Error(`Version ${version} of ${name} already registered.`);
+    }
+    versions.set(version, object);
+
+    // Some pending requests may now be satisfied, so schedule a retry.
+    if (this._pendingRequests.length > 0) {
+      this._debounce.invoke();
+    }
+  }
+
+  /**
+   * Get an object for a name and semVer range.
+   *
+   * @returns a promise resolving to the object with the maximum version
+   * satisfying the version range if one is registered, or undefined if one is
+   * not registered.
+   *
+   * #### Notes
+   * Resolution may be delayed for a short timeout to allow for some delay in
+   * registering the object.
+   */
+  async get(
+    name: string,
+    semVer: string,
+    isKernelRestore: boolean
+  ): Promise<T | undefined> {
+    const mod = this._getSync(name, semVer);
+    if (mod || this._already_retried.indexOf(name + ':' + semVer)) {
+      return mod;
+    }
+    this._already_retried.push(name + ':' + semVer);
+    const result = new PromiseDelegate<T | undefined>();
+    let timestamp = Date.now() + TIMEOUT;
+    if (isKernelRestore) timestamp += TIMEOUT * 4;
+    this._pendingRequests.push({ name, semVer, result, timestamp });
+    this._debounce.invoke();
+    return result.promise;
+  }
+
+  /**
+   * Synchronously get an object for a name and semVer range.
+   *
+   * @returns the object with the maximum version satisfying the version range
+   * or undefined if one does not exist.
+   */
+  private _getSync(name: string, semVer: string): T | undefined {
+    const versions = this._registry.get(name);
+    if (versions) {
+      const best = maxSatisfying(Array.from(versions.keys()), semVer);
+      if (best !== null) {
+        return versions.get(best);
+      }
+    }
+  }
+
+  /**
+   * Retry resolving pending requests.
+   */
+  private _retry(): void {
+    const now = Date.now();
+    // Try to resolve each pending request successfully.
+    this._pendingRequests.forEach(request => {
+      const { name: key, semVer: semver, result } = request;
+      const mod = this._getSync(key, semver);
+      if (mod) {
+        result.resolve(mod);
+        request.timestamp = null;
+      }
+    });
+
+    // Clear out all requests which were fulfilled or expired.
+    ArrayExt.removeAllWhere(this._pendingRequests, ({ result, timestamp }) => {
+      if (timestamp === null) {
+        // Request fulfilled, so remove the entry.
+        return true;
+      } else if (now > timestamp) {
+        // Request expired, so give up and remove the entry.
+        result.resolve(undefined);
+        return true;
+      }
+      return false;
+    });
+
+    // If we still have pending requests, try again later.
+    if (this._pendingRequests.length > 0) {
+      this._debounce.invoke();
+    }
+  }
+
+  private _already_retried = new Array<string>();
+  private _registry = new Map<string, Map<string, T>>();
+  private _pendingRequests: IRequest<T>[] = [];
+  private _debounce = new Debouncer(() => {
```

Comment:
I don't think we want a debouncer. A debouncer will delay calling the function until TIMEOUT ms have passed, which means you can regularly call a debouncer to indefinitely delay the function call. We really do want a throttle, meaning that no matter how many times we call the retry function, it only executes at most once every TIMEOUT ms (but it *does* indeed execute every TIMEOUT ms or so, unlike the debouncer).

Reply:
If we go with the idea of a `timeout` argument to the get function as suggested above, then the timeout on this throttle essentially controls the resolution of that timeout argument.